lib/providers/portion/PortionProvider.ts

```diff
@@ -0,0 +1,255 @@
+      ? CurrencyAmount.fromRawAmount(quoteCurrency, quoteGasAdjusted)
+      : undefined;
+
+    const requestCurrency = sharedInfo.type === TradeType.EXACT_INPUT ? resolvedTokenIn : resolvedTokenOut;
+    const requestCurrencyAmount = requestCurrency
+      ? CurrencyAmount.fromRawAmount(requestCurrency, sharedInfo.amount.toString())
+      : undefined;
+
+    const portion = await this.getPortionForTokenOut(sharedInfo, tokenOutAmount);
+    const portionAmount = portion.portionAmount;
+
+    const quoteGasAndPortionAdjusted = this.getQuoteGasAndPortionAdjusted(
+      sharedInfo.type,
+      portionAmount,
+      quoteCurrencyAmount,
+      quoteGasAdjustedCurrencyAmount,
+      requestCurrencyAmount
+    );
+    return {
+      portionResponse: getPortionResponse,
+      portionAmount: portionAmount,
+      quoteGasAndPortionAdjusted: quoteGasAndPortionAdjusted,
+      portionAdjustedTokenOutAmount: portion.portionAdjustedTokenOutAmount,
+    } as FullPortionPayload;
+  }
+
+  private async resolveTokenInAndTokenOut(
+    sharedInfo: QuoteRequestInfo
+  ): Promise<[Currency | undefined, Currency | undefined]> {
+    try {
+      // we will need to call token fetcher to resolve the tokenIn and tokenOut
+      // there's no guarantee that the tokenIn and tokenOut are in the token address
+      // also the tokenIn and tokenOut can be native token
+      // portion service only accepts wrapped token address
+      return await Promise.all([
+        this.tokenFetcher.resolveTokenBySymbolOrAddress(sharedInfo.tokenInChainId, sharedInfo.tokenIn),
+        this.tokenFetcher.resolveTokenBySymbolOrAddress(sharedInfo.tokenOutChainId, sharedInfo.tokenOut),
+      ]);
+    } catch (e) {
+      log.error({ e }, 'Failed to resolve tokenIn & tokenOut');
+      metrics.putMetric(`PortionProvider.resolveTokenErr`, 1);
+
+      return [undefined, undefined];
+    }
+  }
+
+  private getPortionAmount(
+    tokenOutCurrencyAmount?: CurrencyAmount<Currency>,
+    portionResponse?: GetPortionResponse,
+    resolvedTokenOut?: Currency
+  ): CurrencyAmount<Currency> | undefined {
+    return portionResponse?.hasPortion && portionResponse.portion && resolvedTokenOut && tokenOutCurrencyAmount
+      ? tokenOutCurrencyAmount.multiply(new Fraction(portionResponse.portion.bips, 10000))
+      : undefined;
+  }
+
+  private getPortionAdjustedTokenOutAmount(
+    tradeType: TradeType,
+    tokenOutCurrencyAmount?: CurrencyAmount<Currency>,
+    portionAmount?: CurrencyAmount<Currency>
+  ): CurrencyAmount<Currency> | undefined {
+    if (tokenOutCurrencyAmount && portionAmount) {
+      switch (tradeType) {
+        case TradeType.EXACT_INPUT:
+          return tokenOutCurrencyAmount.subtract(portionAmount);
+        case TradeType.EXACT_OUTPUT:
+          return tokenOutCurrencyAmount.add(portionAmount);
+        default:
+          // unknown trade type.
+          // instead of throw, just return original tokenOutAmount
+          return tokenOutCurrencyAmount;
```

Comment:
I think it's safe to throw here? If we're passing in tradeType from `sharedInfo` then we can be nearly certain that this error case won't happen, and I think it would be really hard to debug if we handle this case silently as is.

Reply:
++ prefer throwing here